stdlib/bool.mli

```diff
@@ -0,0 +1,65 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*                         The OCaml programmers                          *)
+(*                                                                        *)
+(*   Copyright 2018 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Boolean values.
+
+    @since 4.08 *)
+
+(** {1:bools Booleans} *)
+
+type t = bool
+(** The type for boolean values. *)
+
+val not : bool -> bool
+(** [not b] is the boolean negation of [b]. *)
+
+val negate : ('a -> bool) -> ('a -> bool)
+(** [negate p v] is [not (p v)]. *)
+
+external ( && ) : bool -> bool -> bool = "%sequand"
```

Comment:
Why did you choose to export &&/|| as externals, but not equal/compare?

I'm not sure which one to prefer (externals or not).  I'd say that **if** JS backends are clever enough to produce the same efficient code, it's fine to drop external (even if this degrade the bytecode case a bit), otherwise it's a more tricky decision.

Reply:
> Why did you choose to export &&/|| as externals, but not equal/compare?

Because IIRC my tests you don't get the lazy evaluation behaviour if you don't.

Reply:
Ah, right, of course.

Did you check that non-externals equal/compare are correctly optimized for native code (resp. js_of_ocaml)?

Reply:
No. What exactly needs to be checked ? That polymorphic equality is not being used ? That should be the case given the constraints in the implementation.

Reply:
Specialization of primitives occurs before inlining, so I believe it would be broken (a quick test suggests it is indeed the case), even for native code.

Reply:
Ok will switch to externals (side note: maybe this should also be done in the various existing `IntXX` modules).

Reply:
But eta-expansion is performed with the annotations in the `.ml` file (not the .mli I think?) before specialization, so the style used in the implementation, `let compare : bool -> bool -> int = Stdlib.compare` should be fine, and indeed a quick test with `ocaml -dlambda` on 4.07 suggests that it is specialized, as is `equal`.

Reply:
You're right, I did the test without the annotation in the implementation; and this works because Stdlib.compare is exposed as an external (not because of the inliner), which suggests that one must still be careful before removing externals.

Reply:
@dbuenzli so you don't need to use externals here. Only a re-exported polymorphic comparison function would need to be external.